### PR TITLE
[AIRFLOW-3153] send dag last_run to statsd

### PR DESCRIPTION
**Pull request by feng-tao**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3153
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Lyft has been running with this pr for over an year and numerous production issues have been detected by the stats(e.g setting pageduty on the last run time if it exceeds for certain threshold).

This PR adds statds logging for the DAG generation in Airflow, recording

the time spent processing each file; and
the last time it was processed (both as a unix timestamp and as an interval in seconds).

```
Stats.gauge('last_runtime.example_bash_operator', 1.622376)
Stats.gauge('last_run.unixtime.example_bash_operator', 1512670855)
Stats.gauge('last_run.seconds_ago.example_bash_operator', 0.641343)
Stats.gauge('last_runtime.example_bash_operator', 1.629494)
Stats.gauge('last_run.unixtime.example_bash_operator', 1512670886)
Stats.gauge('last_run.seconds_ago.example_bash_operator', 0.526443)
```

Credit to original PR owner(@betodealmeida) at lyft

And fix some flake8 error

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Add stats, no need for test.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by feng-tao** [comment]
PTAL @kaxil

File: airflow/jobs.py [hunk 1]
```diff
@@ -1513,6 +1518,16 @@ def _log_file_processing_stats(self,
                        if processor_start_time else None)
             last_run = processor_manager.get_last_finish_time(file_path)
 
+            file_name = file_path[len(dags_folder) + 1:]
+            dag_name = os.path.splitext(file_name)[0].replace(os.sep, '.')
```
Comment (ashb) [inline airflow/jobs.py:1522]: A dag file could contain multiple DAGs - is there a reason to not use dag_id here?

File: airflow/jobs.py [hunk 2]
```diff
@@ -1513,6 +1518,16 @@ def _log_file_processing_stats(self,
...
+            if last_runtime is not None:
+                Stats.gauge('last_runtime.{}'.format(dag_name), last_runtime)
+            if last_run is not None:
+                unixtime = last_run.strftime("%s")
+                seconds_ago = (timezone.utcnow() - last_run).total_seconds()
+                Stats.gauge('last_run.unixtime.{}'.format(dag_name), unixtime)
+                Stats.gauge('last_run.seconds_ago.{}'.format(dag_name), seconds_ago)
```
Comment (ashb) [inline airflow/jobs.py:1529]: How often does this stat get updated? (I'm not sure from reviewing the PR where we are in the scheduler?) If we were using Promethus I would be tempted to say that just `last_run.unixtime` stat would be the only one we should have, but I honestly' don't remember how Statsd works anymore.

**Review by ashb — COMMENTED** [review]
Stats are good, so I'm not going to complain about more stats, but doesn't Airflow's built in SLA do some of this already?

**Comment by Fokko** [comment]
Tests seem to be failing :'(

**Comment by ashb** [comment]
> pkg_resources.ContextualVersionConflict: (Flask-Login 0.2.11 (/app/.tox/py35-backend_mysql-env_docker/lib/python3.5/site-packages), Requirement.parse('Flask-Login<0.5,>=0.3'), {'flask-appbuilder'})

Rebasing on to latest master should fix that.

**Comment by feng-tao** [comment]
@ashb , @Fokko , sorry for the pr slip through, will update

**Comment by kaxil** [comment]
Can we please add this to docs as well, listing all the stats and what do they mean?

**Comment by andrewhharmon** [comment]
curious why this was implemented as a gauge and not a timer.

**Pull request closed without merging**